nemo/collections/common/tokenizers/tokenizer_utils.py

```diff
@@ -60,38 +61,56 @@ def get_bert_special_tokens(bert_derivative):
+    do_lower_case: Optional[bool] = False,
 ):
     '''
     Args:
     tokenizer_name: sentencepiece or nemobert
-    pretrained_mode_name ('str'): name of the pretrained model, for example: bert-base-cased
-    tokenizer_model (path): only used for sentencepiece tokenizer
-    special_tokens (dict): dict of special tokens (Optional)
-    vocab_file (str): path to vocab file
-    do_lower_case (bool): (whether to apply lower cased) - only applicable when tokenizer is build with vocab file
+    data_file: data file used to build sentencepiece
+    tokenizer_file: tokenizer model file of sentencepiece
+    sample_size: sample size for building sentencepiece
+    pretrained_model_name: name of the pretrained model from the hugging face list or 'megatron',
+        for example: bert-base-cased
+        To see the list of pretrained models, use: nemo_nlp.nm.trainables.get_bert_models_list()
+    tokenizer_model: only used for sentencepiece tokenizer
+    special_tokens: dict of special tokens (Optional)
+    vocab_file: path to vocab file
+    vocab_size: vocab size for building sentence piece
+    do_lower_case: (whether to apply lower cased) - only applicable when tokenizer is build with vocab file or with sentencepiece
     '''
+    # Check if we can use Megatron utils.
+    if pretrained_model_name:
+        model_type = pretrained_model_name.split('-')[0]
+
     if tokenizer_name == 'nemobert':
-        tokenizer = nemo.collections.nlp.data.tokenizers.NemoBertTokenizer(
+        tokenizer = nemo.collections.common.tokenizers.bert_tokenizer.NemoBertTokenizer(
             pretrained_model=pretrained_model_name, vocab_file=vocab_file, do_lower_case=do_lower_case
         )
     elif tokenizer_name == 'sentencepiece':
-        if not os.path.exists(tokenizer_model):
-            raise FileNotFoundError(f'{tokenizer_model} tokenizer model not found')
-
-        tokenizer = nemo.collections.nlp.data.tokenizers.SentencePieceTokenizer(model_path=tokenizer_model)
-        model_type = pretrained_model_name.split('-')[0]
-        if special_tokens is None:
-            if model_type not in MODEL_SPECIAL_TOKENS:
-                logging.info(f'No special tokens found for {model_type}.')
-            else:
+        if not tokenizer_file and not data_file:
+            raise ValueError(f'either tokenizer model or data_file must passed')
+        if not tokenizer_file or not os.path.exists(tokenizer_file):
+            if not special_tokens and pretrained_model_name:
                 special_tokens = MODEL_SPECIAL_TOKENS[model_type]
-        tokenizer.add_special_tokens(special_tokens)
+                num_special_tokens = len(set(special_tokens.values()))
+            tokenizer_file, _ = nemo.collections.common.tokenizers.sentencepiece_tokenizer.create_spt_model(
+                data_file=data_file,
+                vocab_size=vocab_size - num_special_tokens,
+                special_tokens=None,
+                sample_size=sample_size,
```

Comment:
Should it be special_tokens=special_tokens?

Reply:
no, its correct the way it was

Reply:
So one can add special tokens when one trains SPT and when init SPT, will adding special tokens twice break anything?

Reply:
no it will not break anything, but the ultimate vocabulary will be larger with 2x special tokens